### PR TITLE
9.2.4 compatibility

### PR DESCRIPTION
**Pull request by angerman**
VERY WIP

Depends on https://github.com/input-output-hk/cardano-ledger/pull/3098.

This is mostly here to get Hydra building the shell in the background for easier iteration. Then we need to actually push the changes to this branch to get things compiling with 9.2.

----

This moves the ledger to build primarily with GHC 9.2. I say "primarily", because there annoyingly remains one package that doesn't build with 9.2, namely `plutus-preprocessor`. This is because we haven't fixed up `plutus-tx-plugin` to work with 9.2 yet. 

The compromise position I've taken is to move the default build environment to 9.2, but leave instructions for how to get an 8.10 shell if you need to use `plutus-preprocessor`. This requires us to not break 8.10 support for a while, but hopefully not too long. And it still lets us progress towards a 9.2 node.

The alternative would be to just not try to keep 8.10 support at all, and not try to keep `plutus-preprocessor` building. Then we would resurrect it once we fix `plutus-tx-plugin`. I think that's worth considering, since it's not used that frequently and it's not that much code. This approach would allow you to start using 9.2 features everywhere (and give you a simpler CI), but at the cost of losing `plutus-preprocessor` for a bit. Up to you.

### CODE REVIEW

**Comment by JaredCorduan** [comment]
@angerman @erikd can we close the following PRs in favor of this one?

* #3081 
* #3038
* #3033

File: cabal.project [hunk 1]
```diff
@@ -94,3 +97,16 @@ source-repository-package
...
+  location: https://github.com/input-output-hk/plutus
+  tag: cec1eaca9ade859797acec66d9741c9aaac48b45
+  --sha256: 084yvxcx56i5nync597029v23idgvcxysvk7zghsiqfvv4qr911x
+  subdir:
+    plutus-ledger-api
+    plutus-tx
+    plutus-tx-plugin
+    plutus-core
+    prettyprinter-configurable
+    word-array
```
Comment (angerman) [inline cabal.project]: This one should be gone, after @michaelpj's commit, I think

File: eras/alonzo/impl/cardano-ledger-alonzo.cabal [hunk 2]
```diff
@@ -80,9 +80,8 @@ library
     mtl,
     microlens,
     nothunks,
-    plutus-ledger-api ^>= 1.1,
-    plutus-tx ^>= 1.1,
-    plutus-core ^>= 1.1,
+    plutus-ledger-api,
+    plutus-core ^>= 1.0,
```
Comment (angerman) [inline eras/alonzo/impl/cardano-ledger-alonzo.cabal]: This one too. I think you'll be pulling in the wrong plutus-core now.

File: eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal [hunk 3]
```diff
@@ -64,9 +64,8 @@ library
     containers,
     data-default-class,
     hashable >= 1.4.1.0,
-    plutus-core,
     plutus-tx,
-    plutus-ledger-api:{plutus-ledger-api, plutus-ledger-api-testlib} ^>=1.1,
+    plutus-ledger-api,
```
Comment (angerman) [inline eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal]: I'm not sure this is correct?

File: eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal [hunk 4]
```diff
@@ -111,8 +110,7 @@ test-suite cardano-ledger-alonzo-test
     cborg,
     containers,
     data-default-class,
-    plutus-core,
-    plutus-ledger-api:{plutus-ledger-api, plutus-ledger-api-testlib} ^>=1.1,
+    plutus-ledger-api,
```
Comment (angerman) [inline eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal]: neither here

File: eras/babbage/impl/cardano-ledger-babbage.cabal [hunk 5]
```diff
@@ -68,8 +68,7 @@ library
                 deepseq,
                 microlens,
                 nothunks,
-                plutus-ledger-api ^>= 1.1,
-                plutus-tx ^>= 1.1,
+                plutus-ledger-api,
```
Comment (angerman) [inline eras/babbage/impl/cardano-ledger-babbage.cabal]: This also seems to revert Michaels commit.

File: libs/cardano-ledger-test/cardano-ledger-test.cabal [hunk 6]
```diff
@@ -93,7 +93,7 @@ library
     microlens,
     mtl,
     nothunks,
-    plutus-ledger-api:{plutus-ledger-api, plutus-ledger-api-testlib} ^>=1.1,
+    plutus-ledger-api,
```
Comment (angerman) [inline libs/cardano-ledger-test/cardano-ledger-test.cabal]: This one seems to revert Michaels changes as well.

**Comment by angerman** [comment]
I added back @michaelpj's original commentary.

File: eras/babbage/impl/cardano-ledger-babbage.cabal [hunk 7]
```diff
@@ -69,7 +69,7 @@ library
                 microlens,
                 nothunks,
                 plutus-ledger-api ^>= 1.1,
-                plutus-tx ^>= 1.1,
+                -- plutus-tx ^>= 1.1,
```
Comment (michaelpj) [inline eras/babbage/impl/cardano-ledger-babbage.cabal]: commented-out

File: eras/byron/crypto/cardano-crypto-wrapper.cabal [hunk 8]
```diff
@@ -134,7 +132,7 @@ test-suite test
                      , filepath
                      , hedgehog >= 1.0.4
                      , memory
-                     , text
+                     -- , text
```
Comment (michaelpj) [inline eras/byron/crypto/cardano-crypto-wrapper.cabal]: commented-out

**Review by michaelpj — APPROVED** [review]
Looks good!

**Comment by angerman** [comment]
Thanks @michaelpj!

**Comment by angerman** [comment]
The next iteration on this repo, should cut the nix code aggressively for a stock flake if possible.

**Comment by michaelpj** [comment]
Some test failures, looks like it's looking for test files in the wrong place.

**Comment by michaelpj** [comment]
I pushed the haskell.nix bump that fixes the datadir problem.

**Comment by michaelpj** [comment]
The cicero job is broken due to https://input-output.atlassian.net/browse/CIC-246, but Hydra is green. In fact we're about to release a 9.2-compatible plugin, so if we wait a short time then we'll be able to avoid the conditionals an everything.

**Comment by michaelpj** [comment]
Went green!

**Comment by michaelpj** [comment]
@JaredCorduan I think this is good to go now!

**Review by lehins — APPROVED** [review]
:+1: 

Github Actions CI is still failing due to constraint resolution, other than that it looks good

**Review by JaredCorduan — APPROVED** [review]
very exciting, thank you so much to everyone who helped with this!

**Comment by michaelpj** [comment]
Ah indeed, we should mark `plutus-preprocessor` as not buildable on older GHC versions, so cabal doesn't think it should be.

**Comment by michaelpj** [comment]
:tada:

**Pull request merged**